MozziGuts.cpp

```diff
@@ -68,7 +67,7 @@ CircularBuffer<AudioOutput_t> output_buffer;  // fixed size 256
 #  define canBufferAudioOutput() (!output_buffer.isFull())
 #  define bufferAudioOutput(f) output_buffer.write(f)
 static void CACHED_FUNCTION_ATTR defaultAudioOutput() {
-#  if (USE_AUDIO_INPUT == true)
+#  if (USE_AUDIO_INPUT == true && !BYPASS_MOZZI_INPUT_BUFFER == true)
```

Comment:
Maybe, will try to find a good way around this…

Reply:
After some thought, I'd propose more specifically:

In MozziGuts.cpp

```
#define AUDIO_INPUT_NONE 0
#define AUDIO_INPUT_LEGACY 1
#define AUDIO_INPUT_CUSTOM 2

/* ...
[include implementation files; therein, for Mbed:
#if defined(USE_AUDIO_INPUT)
  #define AUDIO_INPUT_MODE AUDIO_INPUT_CUSTOM
#endif
]
... */

#if defined(USE_AUDIO_INPUT) && !defined(AUDIO_INPUT_MODE))
    #define AUDIO_INPUT_MODE AUDIO_INPUT_LEGACY
#elif !defined(AUDIO_INPUT_MODE)
    #define AUDIO_INPUT_MODE AUDIO_INPUT_NONE
#endif  
```

Admittedly, that's quite a large batch of pre-processor ugliness, but at least mostly just in one place. From there on, all #ifs would check for specific named values of AUDIO_INPUT_MODE.

Reply:
But what would happen for a custom implementation that would need the input buffer for instance?

I did not try to amend my initial version for now. As audio_inputs and `mozziAnalogRead` are very very coupled, I think the implementation of the analogReads for MBED (and probably other in the future) will also need to step away for the "legacy" logic of Mozzi. But, the comportment of these two is linked. Hence I wonder if it would not be more strategic to wait for an implementation of `mozziAnalogRead` for MBED and then see if we can separate the two main comportments (ISR of end of conversion and self managed continuous conversions) more globally.

Reply:
> But what would happen for a custom implementation that would need the input buffer for instance?

It could simply define one by itself. The input buffer is not public API in any way (and as we found out, it is probably broken), I see it as an implementation detail of the "legacy" audio input. Similarly, audio input is tightly coupled with mozziAnalogRead(), indeed, in "legacy" mode, but as evidenced by your PR, it doesn't have to be.

Note that part of the suggestion is to move BYPASS_MOZZI_INPUT_BUFFER (or whatever equivalent) out of the config file, into a "user: don't touch this" area, at least for the time being.

Anyway, this is just in the hope to make code more readable, so not super important. I do think it's a good idea not to introduce BYPASS_MOZZI_INPUT_BUFFER _in the config file_, yet, however.

Reply:
> It could simply define one by itself.

Good point indeed, I do not why, but for my these buffers are somehow "unmovable", I have to consider more parts of the code as a playground…

As for coherency, I would also move the BYPASS_OUTPUT_BUFFER out of reach for careless hands then.
Also, I forgot but as the CHUNK is the same for both, I was thinking to put a common value in AudioConfigMbeb, how does that sound?